website/programs.py

```diff
@@ -69,22 +70,32 @@ def save_program(user):
             return 'name must be a string', 400
         if not isinstance(body.get('level'), int):
             return 'level must be an integer', 400
+        if not isinstance(body.get('shared'), bool):
+            return 'shared must be a boolean', 400
         if 'adventure_name' in body:
             if not isinstance(body.get('adventure_name'), str):
                 return 'if present, adventure_name must be a string', 400
 
+        error = False
+        try:
+            hedy.transpile(body.get('code'), body.get('level'), g.lang)
+        except:
+            error = True
+            if not body.get('force_save', True):
+                return jsonify({'parse_error': True})
+
         # We check if a program with a name `xyz` exists in the database for the username.
         # It'd be ideal to search by username & program name, but since DynamoDB doesn't allow searching for two indexes at the same time, this would require to create a special index to that effect, which is cumbersome.
         # For now, we bring all existing programs for the user and then search within them for repeated names.
         programs = DATABASE.programs_for_user(user['username']).records
         program_id = uuid.uuid4().hex
-        program_public = 0
+        program_public = body.get('shared')
         overwrite = False
         for program in programs:
             if program['name'] == body['name']:
                 overwrite = True
                 program_id = program['id']
-                program_public = program.get('public', 0)
+                program_public = program.get('public', program_public)
```

Comment:
This line means that I cannot share a program that was only saved but not shared. This is a deviation from the current behavior. Here are the steps:
- save a program (regardless if it is with errors or not)
- after the program is saved, click Save&Share and approve the override
- go to the my-programs page and check that the program is updated but not shared

Reply:
Will take a look!

Reply:
Fixed!